### DISCUSSION TITLE
상속과 조합의 장단점을 비교해주세요.

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#18] [OOP]
글쎄요 멀까요?

**Comment by alstn113** [comment]
상속이란 기존 클래스에 기능을 추가하거나 재정의하여 새로운 클래스를 정의하는 것을 의미한다.
조합이란 기존 클래스를 상속을 통해 확장하는 대신에 필드로 인스턴스를 참조하게 하는 방법이다.

상속의 단점들에는 다음의 것들이 있다.
1. 객체 간의 의존성이 컴파일 타임에 결정되므로 결합도가 높다. 
2. 자식 클래스가 부모 클래스의 구현에 의존하기 때문에 부모 클래스가 변경되거나 결함이 있을 경우 자식 클래스에 영향을 준다.
3. 자식 클래스가 부모 클래스의 메서드를 오버라이딩할 때 부모 클래스의 호출 방법에 영향을 받는 문제가 있다.
4. 불필요한 인터페이스를 자식 클래스에 상속하는 문제가 있다.

하지만 상속은 코드를 재사용할 뿐만 아니라 부모와 자식의 행동이 호환되는 다형적인 계층 구조를 표현할 수 있다는 장점이 있다.

조합은 객체 간의 의존성이 런타임에 결정되므로 결합도가 낮고, 
상속의 문제점인 메서드 오버라이딩의 오동작, 불필요한 인터페이스 상속 등등에 대한 문제점을 해결한다.
하지만 조합은 의존 관계를 파악하기가 힘들다는 단점이 있다.

**Comment by GIVEN53** [comment]
### 상속
장점
1. 코드 재사용성
    - 부모 클래스의 기능을 자식 클래스에서 재사용할 수 있다.
    - 코드의 중복을 줄이고, 개발 시간을 단축할 수 있다.
2. 계층 구조
    - 클래스를 계층적으로 구성할 수 있다. 
    - 코드를 더욱 추상화하고 구조화할 수 있다.
3. 다형성
    - 부모 클래스의 참조 변수를 사용하여 여러 종류의 자식 클래스 인스턴스를 다룰 수 있다.

단점
1. 강한 결합도
    - 부모 클래스와 자식 클래스 간의 강한 결합을 초래한다. 
    - 부모 클래스의 변경이 자식 클래스에 영향을 미칠 수 있다.
2. 단일 상속 제한
    - Java는 단일 상속만 허용되기 때문에, 클래스가 이미 다른 클래스를 확장하고 있는 경우 추가적인 상속을 받을 수 없다.

### 조합
장점
1. 유연성
    - 상속보다 더 유연한 관계를 형성할 수 있다.
    - 객체들 간의 관계를 동적으로 조정할 수 있고, 단일 상속의 제약을 우회할 수 있다.
2. 낮은 결합도
    - 조합은 클래스 간의 느슨한 결합을 유지할 수 있다. 
    - 한 클래스의 변경이 다른 클래스에 영향을 덜 미칠 수 있다.

단점
1. 코드의 복잡성
    - 객체를 더 작은 구성 요소로 나누고, 이들을 조합하여 기능을 구현하기 때문에 코드의 복잡성이 증가할 수 있다.
2. 재사용 어려움
    - 클래스의 구성 요소를 자주 변경해야 할 수 있으며, 이는 클래스의 재사용을 어렵게 할 수 있다.
